[schema] java.lang.NoClassDefFoundError: org/joda/time/ReadablePartial

**Describe the bug**

When using AvroSchema from latest master, it throws following exceptions:

```
java.lang.RuntimeException: java.lang.NoClassDefFoundError: org/joda/time/ReadablePartial

	at org.apache.pulsar.client.internal.ReflectionUtils.catchExceptions(ReflectionUtils.java:36)
	at org.apache.pulsar.client.internal.DefaultImplementation.newAvroSchema(DefaultImplementation.java:212)
	at org.apache.pulsar.client.api.Schema.AVRO(Schema.java:196)
	at io.streamnative.tests.pulsar.schema.CreateDifferentSchemaAfterDeleteSchemaTest.testCreateSchemaAfterDeletion(CreateDifferentSchemaAfterDeleteSchemaTest.java:84)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at io.streamnative.tests.common.framework.SystemTestRunner.invokeTest(SystemTestRunner.java:219)
	at io.streamnative.tests.common.framework.SystemTestRunner.runChild(SystemTestRunner.java:212)
	at io.streamnative.tests.common.framework.SystemTestRunner.runChild(SystemTestRunner.java:52)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.NoClassDefFoundError: org/joda/time/ReadablePartial
	at org.apache.pulsar.client.impl.schema.AvroSchema.<clinit>(AvroSchema.java:60)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.pulsar.client.internal.DefaultImplementation.lambda$newAvroSchema$24(DefaultImplementation.java:214)
	at org.apache.pulsar.client.internal.ReflectionUtils.catchExceptions(ReflectionUtils.java:34)
	... 30 more
Caused by: java.lang.ClassNotFoundException: org.joda.time.ReadablePartial
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 37 more
```

**To Reproduce**

1. Build the latest master.
2. Use the AvroSchema from latest master. It will throws the above exception.

**Expected behavior**

No exception should be thrown.

**Additional context**

I think the problem introduced by #3900. It doesn't shade the joda library properly.

--- a/distribution/server/src/assemble/LICENSE.bin.txt
+++ b/distribution/server/src/assemble/LICENSE.bin.txt
@@ -464,7 +464,8 @@ The Apache Software License, Version 2.0
     - io.kubernetes-client-java-api-2.0.0.jar
     - io.kubernetes-client-java-proto-2.0.0.jar
   * Joda Time
-    - joda-time-joda-time-2.9.3.jar
+    - joda-time-2.10.1.jar
+    - joda-time-joda-time-2.10.1.jar
   * Dropwizard
     - io.dropwizard.metrics-metrics-core-3.1.0.jar
     - io.dropwizard.metrics-metrics-graphite-3.1.0.jar

--- a/pom.xml
+++ b/pom.xml
@@ -980,6 +980,11 @@ flexible messaging model and an intuitive client API.</description>
         <artifactId>elasticsearch-rest-high-level-client</artifactId>
         <version>${elasticsearch.version}</version>
       </dependency>
+      <dependency>
+        <groupId>joda-time</groupId>
+        <artifactId>joda-time</artifactId>
+        <version>${joda.version}</version>
+      </dependency>
     </dependencies>
   </dependencyManagement>
 

--- a/pulsar-client/pom.xml
+++ b/pulsar-client/pom.xml
@@ -116,6 +116,11 @@
       <groupId>com.fasterxml.jackson.module</groupId>
       <artifactId>jackson-module-jsonSchema</artifactId>
     </dependency>
+
+    <dependency>
+      <groupId>joda-time</groupId>
+      <artifactId>joda-time</artifactId>
+    </dependency>
     
     <!-- httpclient-hostname-verification depends on below dependencies  --> 
     <dependency>
@@ -134,12 +139,6 @@
       <groupId>org.apache.pulsar</groupId>
       <artifactId>pulsar-functions-proto</artifactId>
       <version>${project.parent.version}</version>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>joda-time</groupId>
-      <artifactId>joda-time</artifactId>
-      <version>${joda.version}</version>
       <scope>test</scope>
     </dependency>
   </dependencies>

--- a/pulsar-client/pom.xml
+++ b/pulsar-client/pom.xml
@@ -120,6 +120,7 @@
     <dependency>
       <groupId>joda-time</groupId>
       <artifactId>joda-time</artifactId>
+      <scope>provided</scope>
     </dependency>
     
     <!-- httpclient-hostname-verification depends on below dependencies  --> 
@@ -139,6 +140,11 @@
       <groupId>org.apache.pulsar</groupId>
       <artifactId>pulsar-functions-proto</artifactId>
       <version>${project.parent.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>joda-time</groupId>
+      <artifactId>joda-time</artifactId>
       <scope>test</scope>
     </dependency>
   </dependencies>

--- a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java
+++ b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java
@@ -44,27 +44,32 @@ public class AvroSchema<T> extends StructSchema<T> {
 //      https://issues.apache.org/jira/browse/AVRO-1891  bug address explain
 //      fix the avro logical type read and write
     static {
-        ReflectData reflectDataAllowNull = ReflectData.AllowNull.get();
+        try {
+            ReflectData reflectDataAllowNull = ReflectData.AllowNull.get();
 
-        reflectDataAllowNull.addLogicalTypeConversion(new Conversions.DecimalConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.DateConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimeMicrosConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimestampMicrosConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimeMicrosConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampMicrosConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampConversion());
-        reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimeConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new Conversions.DecimalConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.DateConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimeMicrosConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimestampMicrosConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimeMicrosConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampMicrosConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampConversion());
+            reflectDataAllowNull.addLogicalTypeConversion(new TimeConversions.TimeConversion());
 
-        ReflectData reflectDataNotAllowNull = ReflectData.get();
+            ReflectData reflectDataNotAllowNull = ReflectData.get();
 
-        reflectDataNotAllowNull.addLogicalTypeConversion(new Conversions.DecimalConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.DateConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimeMicrosConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimestampMicrosConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimeMicrosConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampMicrosConversion());
-        reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimeConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new Conversions.DecimalConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.DateConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimeMicrosConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.LossyTimestampMicrosConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimeMicrosConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampMicrosConversion());
+            reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimeConversion());
+        } catch (Throwable t) {
+            LOG.warn("Avro logical types are not available. If you are going to use avro logical types, " +
+                    "you can include `joda-time` in your dependency.");
+        }
     }
 
     private AvroSchema(SchemaInfo schemaInfo) {

--- a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java
+++ b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java
@@ -67,8 +67,10 @@ public class AvroSchema<T> extends StructSchema<T> {
             reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimestampMicrosConversion());
             reflectDataNotAllowNull.addLogicalTypeConversion(new TimeConversions.TimeConversion());
         } catch (Throwable t) {
-            LOG.warn("Avro logical types are not available. If you are going to use avro logical types, " +
-                    "you can include `joda-time` in your dependency.");
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Avro logical types are not available. If you are going to use avro logical types, " +
+                        "you can include `joda-time` in your dependency.");
+            }
         }
     }
 

--- a/pulsar-sql/presto-distribution/LICENSE
+++ b/pulsar-sql/presto-distribution/LICENSE
@@ -253,7 +253,7 @@ The Apache Software License, Version 2.0
     - netty-transport-native-epoll-4.1.33.Final-linux-x86_64.jar
     - netty-transport-native-unix-common-4.1.31.Final.jar
  * Joda Time
-    - joda-time-2.9.9.jar
+    - joda-time-2.10.1.jar
  * Jetty
     - http2-client-9.4.11.v20180605.jar
     - http2-common-9.4.11.v20180605.jar

--- a/pulsar-sql/presto-distribution/LICENSE
+++ b/pulsar-sql/presto-distribution/LICENSE
@@ -253,6 +253,7 @@ The Apache Software License, Version 2.0
     - netty-transport-native-epoll-4.1.33.Final-linux-x86_64.jar
     - netty-transport-native-unix-common-4.1.31.Final.jar
  * Joda Time
+    - joda-time-2.9.9.jar
     - joda-time-2.10.1.jar
  * Jetty
     - http2-client-9.4.11.v20180605.jar

--- a/tests/integration/pom.xml
+++ b/tests/integration/pom.xml
@@ -131,6 +131,12 @@
   	  <artifactId>elasticsearch-rest-high-level-client</artifactId>
   	</dependency>
 
+    <dependency>
+      <groupId>joda-time</groupId>
+      <artifactId>joda-time</artifactId>
+      <scope>test</scope>
+    </dependency>
+
   </dependencies>
 
   <build>

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/SchemaTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/SchemaTest.java
@@ -31,9 +31,16 @@ import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.tests.integration.schema.Schemas.Person;
 import org.apache.pulsar.tests.integration.schema.Schemas.PersonConsumeSchema;
 import org.apache.pulsar.tests.integration.schema.Schemas.Student;
+import org.apache.pulsar.tests.integration.schema.Schemas.AvroLogicalType;
 import org.apache.pulsar.tests.integration.suites.PulsarTestSuite;
+import org.joda.time.DateTime;
+import org.joda.time.LocalDate;
+import org.joda.time.LocalTime;
+import org.joda.time.chrono.ISOChronology;
 import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.Test;
+
+import java.math.BigDecimal;
 
 /**
  * Test Pulsar Schema.
@@ -155,6 +162,60 @@ public class SchemaTest extends PulsarTestSuite {
             assertEquals("man", personConsumeSchema.getGender());
             log.info("Successfully consumer personConsumeSchema : {}", personConsumeSchema);
         }
+    }
+
+    @Test
+    public void testAvroLogicalType() throws Exception {
+        final String tenant = PUBLIC_TENANT;
+        final String namespace = "test-namespace-" + randomName(16);
+        final String topic = "test-logical-type-schema";
+        final String fqtn = TopicName.get(
+                TopicDomain.persistent.value(),
+                tenant,
+                namespace,
+                topic
+        ).toString();
+
+        admin.namespaces().createNamespace(
+                tenant + "/" + namespace,
+                Sets.newHashSet(pulsarCluster.getClusterName())
+        );
+
+        AvroLogicalType messageForSend = AvroLogicalType.builder()
+                .decimal(new BigDecimal("12.34"))
+                .timestampMicros(System.currentTimeMillis() * 1000)
+                .timestampMillis(new DateTime("2019-03-26T04:39:58.469Z", ISOChronology.getInstanceUTC()))
+                .timeMillis(LocalTime.now())
+                .timeMicros(System.currentTimeMillis() * 1000)
+                .date(LocalDate.now())
+                .build();
+
+        try (Producer<AvroLogicalType> producer = client
+                .newProducer(Schema.AVRO(AvroLogicalType.class))
+                .topic(fqtn)
+                .create()
+        ) {
+            producer.send(messageForSend);
+            log.info("Successfully published avro logical type message : {}", messageForSend);
+        }
+
+        try (Consumer<AvroLogicalType> consumer = client
+                .newConsumer(Schema.AVRO(AvroLogicalType.class))
+                .topic(fqtn)
+                .subscribe()
+        ) {
+            AvroLogicalType received = consumer.receive().getValue();
+            assertEquals(messageForSend.getDecimal(), received.getDecimal());
+            assertEquals(messageForSend.getTimeMicros(), received.getTimeMicros());
+            assertEquals(messageForSend.getTimeMillis(), received.getTimeMillis());
+            assertEquals(messageForSend.getTimestampMicros(), received.getTimestampMicros());
+            assertEquals(messageForSend.getTimestampMillis(), received.getTimestampMillis());
+            assertEquals(messageForSend.getDate(), received.getDate());
+
+            log.info("Successfully consumer avro logical type message : {}", received);
+        }
+
+
     }
 
 }

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/Schemas.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/Schemas.java
@@ -31,12 +31,18 @@
  */
 package org.apache.pulsar.tests.integration.schema;
 
+import lombok.Builder;
 import lombok.Data;
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.Setter;
 import lombok.ToString;
 import org.apache.avro.reflect.AvroDefault;
+import org.joda.time.DateTime;
+import org.joda.time.LocalDate;
+import org.joda.time.LocalTime;
+
+import java.math.BigDecimal;
 
 /**
  * Keep a list of schemas for testing.
@@ -90,6 +96,30 @@ public final class Schemas {
         private int gpa;
         private int grade;
 
+    }
+
+    @Data
+    @ToString
+    @EqualsAndHashCode
+    @Builder
+    public static class AvroLogicalType{
+        @org.apache.avro.reflect.AvroSchema("{\n" +
+                "  \"type\": \"bytes\",\n" +
+                "  \"logicalType\": \"decimal\",\n" +
+                "  \"precision\": 4,\n" +
+                "  \"scale\": 2\n" +
+                "}")
+        BigDecimal decimal;
+        @org.apache.avro.reflect.AvroSchema("{\"type\":\"int\",\"logicalType\":\"date\"}")
+        LocalDate date;
+        @org.apache.avro.reflect.AvroSchema("{\"type\":\"long\",\"logicalType\":\"timestamp-millis\"}")
+        DateTime timestampMillis;
+        @org.apache.avro.reflect.AvroSchema("{\"type\":\"int\",\"logicalType\":\"time-millis\"}")
+        LocalTime timeMillis;
+        @org.apache.avro.reflect.AvroSchema("{\"type\":\"long\",\"logicalType\":\"timestamp-micros\"}")
+        long timestampMicros;
+        @org.apache.avro.reflect.AvroSchema("{\"type\":\"long\",\"logicalType\":\"time-micros\"}")
+        long timeMicros;
     }
 
     private Schemas() {}

